cloud/services/managedclusters/managedclusters.go

```diff
@@ -0,0 +1,211 @@
+	"net"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2020-02-01/containerservice"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+	azure "sigs.k8s.io/cluster-api-provider-azure/cloud"
+)
+
+var defaultUser string = "azureuser"
+
+// Spec contains properties to create a managed cluster.
+type Spec struct {
+	// Name is the name of this AKS Cluster.
+	Name string
+
+	// ResourceGroup is the name of the Azure resource group for this AKS Cluster.
+	ResourceGroup string
+
+	// Location is a string matching one of the canonical Azure region names. Examples: "westus2", "eastus".
+	Location string
+
+	// Tags is a set of tags to add to this cluster.
+	// +optional
+	Tags map[string]string
+
+	// Version defines the desired Kubernetes version.
+	Version string
+
+	// LoadBalancerSKU for the managed cluster. Possible values include: 'Standard', 'Basic'. Defaults to standard.
+	LoadBalancerSKU *string
+
+	// NetworkPlugin used for building Kubernetes network. Possible values include: 'Azure', 'Kubenet'. Defaults to Azure.
+	NetworkPlugin *string
+
+	// NetworkPolicy used for building Kubernetes network. Possible values include: 'NetworkPolicyCalico', 'NetworkPolicyAzure'. Defaults to Azure.
+	NetworkPolicy *string
+
+	// SSHPublicKey is a string literal containing an ssh public key. Will autogenerate and discard if not provided.
+	SSHPublicKey string
+
+	// AgentPools is the list of agent pool specifications in this cluster.
+	AgentPools []PoolSpec
+
+	// PodCIDR is the CIDR block for IP addresses distributed to pods
+	PodCIDR string
+
+	// ServiceCIDR is the CIDR block for IP addresses distributed to services
+	ServiceCIDR string
+}
+
+type PoolSpec struct {
+	Name         string
+	SKU          string
+	Replicas     int32
+	OSDiskSizeGB int32
+}
+
+// Get fetches a managed cluster from Azure.
+func (s *Service) Get(ctx context.Context, spec interface{}) (interface{}, error) {
+	managedClusterSpec, ok := spec.(*Spec)
+	if !ok {
+		return containerservice.ManagedCluster{}, errors.New("expected managed cluster specification")
+	}
+	return s.Client.Get(ctx, managedClusterSpec.ResourceGroup, managedClusterSpec.Name)
+}
+
+// Get fetches a managed cluster kubeconfig from Azure.
+func (s *Service) GetCredentials(ctx context.Context, spec interface{}) ([]byte, error) {
+	managedClusterSpec, ok := spec.(*Spec)
+	if !ok {
+		return nil, errors.New("expected managed cluster specification")
+	}
+	return s.Client.GetCredentials(ctx, managedClusterSpec.ResourceGroup, managedClusterSpec.Name)
+}
+
+// Reconcile idempotently creates or updates a managed cluster, if possible.
+func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
+	managedClusterSpec, ok := spec.(*Spec)
+	if !ok {
+		return errors.New("expected managed cluster specification")
+	}
+
+	// ****WARNING****
+	// ****WARNING****
+	// TODO(ace): probably don't do this.
+	settings, err := auth.GetSettingsFromEnvironment()
```

Comment:
probably needs to be passed into this function after being fetched from a kubernetes secret? put secret name in the control plane spec?